src/backend/distributed/commands/multi_copy.c

```diff
@@ -1981,6 +1999,44 @@ CreateCitusCopyDestReceiver(Oid tableId, List *columnNameList, int partitionColu
 }
 
 
+/*
+ * ShouldExecuteCopyLocally returns true if the current copy
+ * operation should be done locally for local placements.
+ */
+static bool
+ShouldExecuteCopyLocally()
+{
+	if (!EnableLocalExecution)
+	{
+		return false;
+	}
+
+	if (TransactionAccessedLocalPlacement)
+	{
+		/*
+		 * For various reasons, including the transaction visibility
+		 * rules (e.g., read-your-own-writes), we have to use local
+		 * execution again if it has already happened within this
+		 * transaction block.
+		 *
+		 * We might error out later in the execution if it is not suitable
+		 * to execute the tasks locally.
+		 */
+		Assert(IsMultiStatementTransaction() || InCoordinatedTransaction());
+
+		/*
+		 * TODO: A future improvement could be to keep track of which placements
+		 * have been locally executed. At this point, only use local execution for
+		 * those placements. That'd help to benefit more from parallelism.
+		 */
+
+		return true;
+	}
+
+	return IsTransactionBlock();
```

Comment:
1-> Yeah, we should use `IsMultiStatementTransaction` I have updated it. I will also add some more tests for that
2-> I guess at this line, we are always in a coordinated transaction, because at the beginning of copy we call `UseCoordinatedTransaction`. So checking that would mean we are always going to be running the copy locally.
3-> When we do a local copy, we lose parallelism, so if we are not in a transaction block, it might be better to perform the copy in parallel. We should do a performance test here as well I guess